### PR TITLE
Add HTTP backup store for Raptor	

### PR DESCRIPTION
**Pull request by electrum**

### CODE REVIEW

**Comment by electrum** [comment]
Depends on https://github.com/airlift/airlift/pull/423

File: presto-raptor/src/main/java/com/facebook/presto/raptor/backup/HttpBackupStore.java [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+            throw propagate(request, exception);
+        }
+
+        @Override
+        public StatusResponse handle(Request request, Response response)
+                throws IOException
+        {
+            StatusResponse status = createStatusResponse(response);
+            if (isOk(status)) {
+                writeFile(response.getInputStream());
```
Comment (dain) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/backup/HttpBackupStore.java]: Did you verify the client is not buffering the whole response?

File: presto-raptor/src/test/java/com/facebook/presto/raptor/backup/TestHttpBackupStore.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                .put("backup.http.uri", "http://localhost:8080")
+                .build();
+
+        Bootstrap app = new Bootstrap(
+                new TestingNodeModule(),
+                new TestingHttpServerModule(),
+                new JsonModule(),
+                new JaxrsModule(true),
+                binder -> jaxrsBinder(binder).bind(TestingHttpBackupResource.class),
+                override(new HttpBackupModule()).with(new TestingModule()));
```
Comment (dain) [inline presto-raptor/src/test/java/com/facebook/presto/raptor/backup/TestHttpBackupStore.java]: I didn't think that override worked reliably

File: presto-raptor/src/test/java/com/facebook/presto/raptor/backup/TestingHttpBackupResource.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    @Path("{uuid}")
+    public synchronized Response deleteRequest(
+            @HeaderParam(PRESTO_ENVIRONMENT) String environment,
+            @PathParam("uuid") UUID uuid)
+    {
+        checkEnvironment(environment);
+        if (!shards.containsKey(uuid)) {
+            return Response.status(NOT_FOUND).build();
+        }
+        if (shards.get(uuid) == null) {
```
Comment (dain) [inline presto-raptor/src/test/java/com/facebook/presto/raptor/backup/TestingHttpBackupResource.java]: This is not thread safe.  Instead use:
```
   byte[] oldValue = shards.replace(uuid, null);
   if (oldValue == null) {
```

File: presto-raptor/src/test/java/com/facebook/presto/raptor/backup/TestingHttpBackupResource.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            byte[] bytes)
+    {
+        checkEnvironment(environment);
+        if ((request.getContentLength() < 0) || (bytes.length != request.getContentLength())) {
+            return Response.status(BAD_REQUEST).build();
+        }
+        if (parseUnsignedLong(hexHash, 16) != XxHash64.hash(Slices.wrappedBuffer(bytes))) {
+            return Response.status(BAD_REQUEST).build();
+        }
+        if (shards.containsKey(uuid)) {
```
Comment (dain) [inline presto-raptor/src/test/java/com/facebook/presto/raptor/backup/TestingHttpBackupResource.java]: Use `putIfAbsent`

File: presto-raptor/src/test/java/com/facebook/presto/raptor/backup/TestingHttpBackupResource.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    @Path("{uuid}")
+    public synchronized Response headRequest(
+            @HeaderParam(PRESTO_ENVIRONMENT) String environment,
+            @PathParam("uuid") UUID uuid)
+    {
+        checkEnvironment(environment);
+        if (!shards.containsKey(uuid)) {
+            return Response.status(NOT_FOUND).build();
+        }
+        if (shards.get(uuid) == null) {
```
Comment (dain) [inline presto-raptor/src/test/java/com/facebook/presto/raptor/backup/TestingHttpBackupResource.java]: you could use something like this:
```java
private static final byte[] NOT_PRESENT = new byte[];

byte[] value = shards.getOrDefault(uuid, NOT_PRESENT);
if (value == NOT_PRESENT) {
    return Response.status(NOT_FOUND).build();
}
if (value == null) {
    return Response.status(GONE).build();
}
return Response.noContent().build();
```

File: presto-raptor/src/test/java/com/facebook/presto/raptor/backup/TestingHttpBackupResource.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    public synchronized Response getRequest(
+            @HeaderParam(PRESTO_ENVIRONMENT) String environment,
+            @PathParam("uuid") UUID uuid)
+    {
+        checkEnvironment(environment);
+        if (!shards.containsKey(uuid)) {
+            return Response.status(NOT_FOUND).build();
+        }
+        byte[] bytes = shards.get(uuid);
+        if (bytes == null) {
```
Comment (dain) [inline presto-raptor/src/test/java/com/facebook/presto/raptor/backup/TestingHttpBackupResource.java]: I don't see any place where you insert a null into the map.

File: presto-raptor/src/test/java/com/facebook/presto/raptor/backup/TestingHttpBackupResource.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+            @PathParam("uuid") UUID uuid)
+    {
+        checkEnvironment(environment);
+        if (!shards.containsKey(uuid)) {
+            return Response.status(NOT_FOUND).build();
+        }
+        if (shards.get(uuid) == null) {
+            return Response.status(GONE).build();
+        }
+        shards.remove(uuid);
```
Comment (electrum) [inline presto-raptor/src/test/java/com/facebook/presto/raptor/backup/TestingHttpBackupResource.java]: This should add a null

File: presto-raptor/src/test/java/com/facebook/presto/raptor/backup/TestHttpBackupStore.java (hunk 2 above)
Comment (electrum) [inline presto-raptor/src/test/java/com/facebook/presto/raptor/backup/TestHttpBackupStore.java]: Seems to work fine here

**Pull request closed without merging**